### PR TITLE
Device(smart-NICs) Plugin: Fix host networking limitation

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:
With the current [Device Plugin](https://github.com/kubernetes/kubernetes/pull/51209) changes, if any smart NIC device plugin is used, there is a limitation to use host-networking for the pod containers to see the network interface inside containers.
Pass namespace, pod name, container name(along with device IDs) in the gRPC `Allocate`
api to the remote device plugin.
More context [here](https://groups.google.com/forum/#!topic/kubernetes-wg-resource-management/t9LMWgU8tXI)
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Here is how these changes can be used by device plugin:
https://github.com/vikaschoudhary16/kubernetes/pull/7
**Release note**:

```release-note
```
/cc @sjenning @jeremyeder

### CODE REVIEW

**Comment by jeremyeder** [comment]
@sjenning  can you add area/hw-accelerators label?  How can I add labels?

**Comment by sjenning** [comment]
@vikaschoudhary16 @jeremyeder what is the thinking for this PR?  It makes changes to the device plugin API that have not been agreed upon.  So 1.9, pending agreement from @RenaudWasTaken and others?

It isn't a "fix" as much as an expansion of the API to allow for this networking use-case.

**Comment by vikaschoudhary16** [comment]
@sjenning My intention was to initiate a discussion with this PR. So that it could be easier to explain what I am suggesting. Now looking forward to the community views.

**Comment by jeremyeder** [comment]
@dcbw

**Comment by tallclair** [comment]
@kubernetes/sig-network-pr-reviews

File: pkg/kubelet/apis/deviceplugin/v1alpha1/api.proto [hunk 1]
```diff
@@ -78,7 +78,11 @@ message Device {
 // - Allocate allows Device Plugin to run device specific operations on
 //   the Devices requested
 message AllocateRequest {
-	repeated string devicesIDs = 1;
+        // Namespace to which the container requesting device allocation belongs
+        string namespace = 1;
+        // Name of the pod to which container requesting device allocation belongs
+        string podName = 2;
+	repeated string devicesIDs = 3;
```
Comment (feiskyer) [inline pkg/kubelet/apis/deviceplugin/v1alpha1/api.proto]: nit: Add comments for devicesIDs and keep order of those fields so it won't break old APIs.

**Comment by fabiand** [comment]
I do like the approach of reusing the device plugin mechanism to get network interfaces into a pod as well.

The only drawback of this PR is that it requires the NIC to be backed by a physical device. If this could be softend, then this approach could also be used for virtual NICs which could be backed by SDNs, vlans or alike.

**Pull request closed without merging**